imports/plugins/core/discounts/client/components/form.js

```diff
@@ -0,0 +1,127 @@
+import React, { Component, PropTypes } from "react";
+import { Meteor } from "meteor/meteor";
+import debounce from "lodash/debounce";
+import { Translation } from "/imports/plugins/core/ui/client/components";
+
+export default class DiscountForm extends Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      discount: this.props.discount,
+      validatedInput: false,
+      attempts: 0,
+      discountApplied: false
+    };
+
+    // debounce helper so to wait on user input
+    this.debounceDiscounts = debounce(() => {
+      const { discount } = this.state;
+      Meteor.call("discounts/codes/apply", this.props.id, discount, this.props.collection, (error, results) => {
+        if (results) {
+          this.setState({ discountApplied: results, discount: "" });
+        } else {
+          this.setState({ discountApplied: false });
+        }
+      });
+    }, 500);
+
+    this.handleChange = this.handleChange.bind(this);
+    this.handleClick = this.handleClick.bind(this);
+  }
+
+  // handle apply
+  renderApplied() {
+    return (
+      <a onClick={this.handleClick}>
+        <Translation defaultValue="Discount submitted." i18nKey="discounts.submitted"/>
+      </a>
+    );
+  }
+
+  // handle keydown and change events
+  handleChange(event) {
+    const { attempts } = this.state;
+    // clear input if user hits escape key
+    if (event.keyCode === 27) {
+      return this.setState({
+        discount: "",
+        validatedInput: false,
+        attempts: 0,
+        discountApplied: false
+      });
+    }
+    this.setState({ discount: event.target.value, attempts: attempts + 1 });
+    // TODO: this.debounce doesn't always need to exec
+    // we should add some logic to determine based on attempts
+    // or some other cleverness if now is a good time to apply the code.
+    return this.debounceDiscounts();
+  }
+
+  // handle display or not
+  handleClick(event) {
+    event.preventDefault();
+    this.setState({ validatedInput: true });
+  }
+
+  // loader button
+  loader() {
+    const { attempts, discount } = this.state;
+    let loader;
+    if (discount && discount.length >= 10 && attempts >= 12) {
```

Comment:
yes, but I was going to leave the details of this to another UI task.  Right now this doesn't throttle but, does show a different icon as it becomes more unlikely that a match will be made.   I wasn't sure exactly what rules we'd want to enforce.. a max length, or number of bad attempts past some character count...? Open to ideas..